pyo3-macros-backend/src/pyimpl.rs

```diff
@@ -84,6 +85,18 @@ pub fn build_py_methods(
     }
 }
 
+fn check_pyfunction(meth: &mut ImplItemFn) -> syn::Result<()> {
+    if meth
+        .attrs
+        .iter()
+        .any(|attr| attr.path().is_ident("pyfunction"))
+    {
+        meth.attrs.clear();
```

Comment:
IMO `.clear()` is a bit too much because it might introduce other compile errors by removing needed attributes.

One option is to use `.retain()`, something like this pseudocode:

```rust
let mut error = None;
meth.attrs.retain(|attr| {
    if attr is pyfunction { 
        error = Some(err_spanned!(...));
        false
    } else {
        true
    }
});
error.map_or(Ok(()), Err)

Reply:
Thanks for the pseudo code, it helped a lot!

I make a copy of attrs to construct a 1-element slice here:

```
        let attrs = [attr.clone()];
```

idk what is the idiomatic way in rust to solve this issue ,but this is the best I could do.

There could be two additional function for `has_attribute_with_namespace` and `has_attribute` accepting just a single attribute but I felt it is easier to construct a one-element slice here.

Span needs to be adjusted a little bit as in retain the attrs is not yet removed at that point so rust points to the attribute instead if I use `meth.span()`, like this:

```
error: functions inside #[pymethods] do not need to be annotated with #[pyfunction]
  --> src/lib.rs:10:9
   |
10 |         #[pyfunction]
   |         ^
```